File: src/main/java/pranav/views/FloatingMenu/FMHelper.java
package pranav.views.FloatingMenu;

import android.content.Context;
import android.graphics.Color;
import android.graphics.drawable.ColorDrawable;
import android.graphics.drawable.Drawable;
import androidx.annotation.ColorInt;
import androidx.annotation.DrawableRes;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.annotation.StringRes;
import androidx.appcompat.app.AppCompatActivity;
import android.view.View.OnClickListener;
import android.view.Window;

import com.preons.pranav.utilities.R;

import java.util.Arrays;

import pranav.utilities.Utilities;

import static pranav.utilities.Utilities.Resources.getColoredDrawable;


/**
 * Created on 31-05-17 at 12:01 AM by Pranav Raut.
 * For QRCodeProtection
 */

@SuppressWarnings({"WeakerAccess", "unused"})
public final class FMHelper {

    private int numOptions;
    private OnClickListener[] listeners;
    private Drawable mainBtnRes;
    private Drawable[] optionBtnRes;
    @ColorInt
    private int statusBarColor = 0x55000000;
    private Drawable imageBackground = new ColorDrawable(statusBarColor);
    @ColorInt
    private int mainBtnBaseColor;
    private int mainBtnOpenColor = 0xffffffff;
    @ColorInt
    private int[] optionBtnColors;
    private String optionTexts[];
    private Drawable optionTextBackground;
    private Utilities.Resources resources;
    private boolean isAnimatedStatusBar = true;
    private boolean isTintedFabSrc = true;
    private boolean isTintedStatusBarIcon = true;
    private boolean useCard = true;
    private Window window;

    public FMHelper(Context context) {
        this(3, context);
    }

    public FMHelper(int numOptions, Context context) {
        this.numOptions = numOptions;
        window = ((AppCompatActivity) context).getWindow();
        resources = new Utilities.Resources(context);
        optionBtnRes = new Drawable[numOptions];
        optionBtnColors = new int[numOptions];
        optionTexts = new String[numOptions];
        listeners = new OnClickListener[numOptions];
        mainBtnRes = resources.getDrawable(R.drawable.ic_add);
        Arrays.fill(optionBtnColors, mainBtnBaseColor = resources.getColor(R.color.colorPrimary));
        Arrays.fill(optionBtnRes, resources.getDrawable(R.drawable.ic_add));
        for (int i = 0; i < numOptions; i++) optionTexts[i] = "Option #" + (i + 1);
        this.optionTextBackground = getColoredDrawable(resources.getDimen(R.dimen.pad4dp), 0x90ffffff);
    }


    public int getNumOptions() {
        return numOptions;
    }

    public FMHelper setNumOptions(int numOptions) {
        this.numOptions = numOptions;
        return this;
    }

    public Drawable getMainBtnRes() {
        return mainBtnRes;
    }

    public FMHelper setMainBtnRes(Drawable mainBtnRes) {
        this.mainBtnRes = mainBtnRes;
        return this;
    }

    public FMHelper setMainBtnRes(@DrawableRes int mainBtnRes) {
        return setMainBtnRes(resources.getDrawable(mainBtnRes));
    }

    public OnClickListener[] getListeners() {
        return listeners;
    }

    public FMHelper setListeners(OnClickListener... listeners) {
        if (listeners.length < numOptions)
            throw new IllegalArgumentException("Length of listeners given is not equal to numOptions set in constructor");
        this.listeners = listeners;
        return this;
    }

    public Drawable[] getOptionBtnRes() {
        return optionBtnRes;
    }

    public FMHelper setOptionBtnRes(@DrawableRes int[] optionBtnRes) {
        if (optionBtnRes.length < numOptions)
            throw new IllegalArgumentException("Length of drawables given is not equal to numOptions set in constructor");
        for (int i = 0; i < numOptions; i++)
            this.optionBtnRes[i] = resources.getDrawable(optionBtnRes[i]);
        return this;
    }

    public FMHelper setOptionBtnRes(String[] resName) {
        if (resName.length < numOptions)
            throw new IllegalArgumentException("Length of drawables given is not equal to numOptions set in constructor ("
                    + numOptions + ", " + optionBtnColors.length + ")");
        for (int i = 0; i < numOptions; i++)
            this.optionBtnRes[i] = resources.getDrawable(resName[i]);
        return this;
    }

    public void setOptionBtnRes(Drawable[] optionBtnRes) {
        if (optionBtnRes.length < numOptions)
            throw new IllegalArgumentException("Length of drawables given is not equal to numOptions set in constructor");
        this.optionBtnRes = optionBtnRes;
    }

    public Drawable getImageBackground() {
        return imageBackground;
    }

    public FMHelper setImageBackground(Drawable imageBackground) {
        this.imageBackground = imageBackground;
        return this;
    }

    public FMHelper setImageBackground(@ColorInt int backColor) {
        this.imageBackground = new ColorDrawable(backColor);
        return this;
    }

    @ColorInt
    public int getMainBtnBaseColor() {
        return mainBtnBaseColor;
    }

    public FMHelper setMainBtnBaseColor(@ColorInt int color) {
        this.mainBtnBaseColor = color;
        return this;
    }

    public FMHelper setMainBtnBaseColor(String color) {
        return setMainBtnBaseColor(Color.parseColor(color));
    }

    @ColorInt
    public int getMainBtnOpenColor() {
        return mainBtnOpenColor;
    }

    public FMHelper setMainBtnOpenColor(@ColorInt int color) {
        this.mainBtnOpenColor = color;
        return this;
    }

    public FMHelper setMainBtnOpenColor(String color) {
        return setMainBtnOpenColor(Color.parseColor(color));
    }

    @ColorInt
    public int[] getOptionBtnColors() {
        return optionBtnColors;
    }

    public FMHelper setOptionBtnColors(@ColorInt int[] optionBtnColors) {
        if (optionBtnColors.length < numOptions)
            throw new IllegalArgumentException("Length of colors given is not equal to numOptions set in constructor ("
                    + numOptions + ", " + optionBtnColors.length + ")");
        this.optionBtnColors = optionBtnColors;
        return this;
    }

    public FMHelper setOptionBtnColors(@ColorInt int optionBtnColor) {
        Arrays.fill(optionBtnColors, optionBtnColor);
        return this;
    }

    public String[] getOptionTexts() {
        return optionTexts;
    }

    public FMHelper setOptionTexts(String[] optionTexts) {
        if (optionTexts.length < numOptions)
            throw new IllegalArgumentException("Length of texts given is not equal to numOptions set in constructor ("
                    + numOptions + ", " + optionTexts.length + ")");
        this.optionTexts = optionTexts;
        return this;
    }

    public FMHelper setOptionTexts(@StringRes int[] optionTexts) {
        if (optionTexts.length != numOptions)
            throw new IllegalArgumentException("Length of texts given is not equal to numOptions set in constructor ("
                    + numOptions + ", " + optionTexts.length + ")");
        for (int i = 0; i < optionTexts.length; i++)
            this.optionTexts[i] = resources.getString(optionTexts[i]);
        return this;
    }

    public Drawable getOptionTextBackground() {
        return optionTextBackground;
    }

    public FMHelper setOptionTextBackground(@DrawableRes int optionTextBackground) {
        return setOptionTextBackground(resources.getDrawable(optionTextBackground));
    }

    public FMHelper setOptionTextBackground(Drawable optionTextBackground) {
        this.optionTextBackground = optionTextBackground;
        return this;
    }

    public FMHelper setOptionTextBackgroundColor(@ColorInt int color) {
        return setOptionTextBackground(new ColorDrawable(color));
    }

    public int getStatusBarColor() {
        return statusBarColor;
    }

    public FMHelper setStatusBarColor(@ColorInt int statusBarColor) {
        this.statusBarColor = statusBarColor;
        return this;
    }

    public FMHelper setStatusBarColor(String color) {
        return setStatusBarColor(Color.parseColor(color));
    }

    public boolean isAnimatedStatusBar() {
        return isAnimatedStatusBar;
    }

    public FMHelper setAnimatedStatusBar(boolean animatedStatusBar) {
        this.isAnimatedStatusBar = animatedStatusBar;
        return this;
    }

    public boolean isTintedFabSrc() {
        return isTintedFabSrc;
    }

    public FMHelper setTintedFabSrc(boolean tintedFabSrc) {
        isTintedFabSrc = tintedFabSrc;
        return this;
    }

    @Nullable
    public Window getWindow() {
        return window;
    }

    public FMHelper setWindow(@NonNull Window window) {
        this.window = window;
        return this;
    }

    boolean getUseCard() {
        return useCard;
    }

    FMHelper setUseCard(boolean useCard) {
        this.useCard = useCard;
        return this;
    }
}

File: src/main/java/pranav/views/floatingMenu/FMHelper.kt
package pranav.views.floatingMenu

import android.content.Context
import android.graphics.Color
import android.graphics.drawable.ColorDrawable
import android.graphics.drawable.Drawable
import android.view.View.OnClickListener
import android.view.Window
import androidx.annotation.ColorInt
import androidx.annotation.DrawableRes
import androidx.annotation.StringRes
import androidx.appcompat.app.AppCompatActivity
import com.preons.pranav.utilities.R
import pranav.utilities.Utilities
import pranav.utilities.Utilities.ResourceManager.Companion.getColoredDrawable
import java.util.*


/**
 * Created on 31-05-17 at 12:01 AM by Pranav Raut.
 * For QRCodeProtection
 */

class FMHelper(var numOptions: Int, context: Context) {
    var listeners: Array<OnClickListener>? = null
    var mainBtnRes: Drawable? = null
    var optionBtnRes: Array<Drawable>? = null
    @ColorInt
    var statusBarColor = 0x55000000
    var imageBackground: Drawable = ColorDrawable(statusBarColor)
    @ColorInt
    var mainBtnBaseColor: Int = 0
    var mainBtnOpenColor = -0x1
    @ColorInt
    var optionBtnColors: Array<Int>? = null
    var optionTexts: Array<String>? = null
    var optionTextBackground: Drawable? = null
    private val resourceManager: Utilities.ResourceManager
    var isAnimatedStatusBar = true
    var isTintedFabSrc = true
    private val isTintedStatusBarIcon = true
    private var useCard = true
    var window: Window? = null

    constructor(context: Context) : this(3, context) {}

    init {
        window = (context as AppCompatActivity).window
        resourceManager = Utilities.ResourceManager(context)
        optionBtnRes = emptyArray()
        optionBtnColors = emptyArray()
        optionTexts = emptyArray()
        listeners = emptyArray()
        mainBtnRes = resourceManager.getDrawable(R.drawable.ic_add)
        mainBtnBaseColor = resourceManager.getColor(R.color.colorPrimary)

        Arrays.fill(optionBtnColors, mainBtnBaseColor)
        Arrays.fill(optionBtnRes, resourceManager.getDrawable(R.drawable.ic_add))
        for (i in 0 until numOptions) optionTexts!![i] = "Option #" + (i + 1)
        this.optionTextBackground =
            getColoredDrawable(resourceManager.getDimen(R.dimen.pad4dp), -0x6f000001)
    }


    fun setNumOptions(numOptions: Int): FMHelper {
        this.numOptions = numOptions
        return this
    }


    fun setMainBtnRes(mainBtnRes: Drawable?): FMHelper {
        this.mainBtnRes = mainBtnRes
        return this
    }

    fun setMainBtnRes(@DrawableRes mainBtnRes: Int): FMHelper {
        return setMainBtnRes(resourceManager.getDrawable(mainBtnRes))
    }


    fun setListeners(vararg listeners: OnClickListener): FMHelper {
        require(listeners.size >= numOptions) { "Length of listeners given is not equal to numOptions set in constructor" }
        this.listeners = arrayOf(*listeners)
        return this
    }

    fun setOptionBtnRes(@DrawableRes optionBtnRes: IntArray): FMHelper {
        require(optionBtnRes.size >= numOptions) { "Length of drawables given is not equal to numOptions set in constructor" }
        for (i in 0 until numOptions)
            this.optionBtnRes?.set(i, resourceManager.getDrawable(optionBtnRes[i])!!)
        return this
    }

    fun setOptionBtnRes(resName: Array<String>): FMHelper {
        require(resName.size >= numOptions) {
            ("Length of drawables given is not equal to numOptions set in constructor ("
                    + numOptions + ", " + optionBtnColors!!.size + ")")
        }
        for (i in 0 until numOptions)
            this.optionBtnRes?.set(i, resourceManager.getDrawable(resName[i])!!)
        return this
    }


    fun setImageBackground(imageBackground: Drawable): FMHelper {
        this.imageBackground = imageBackground
        return this
    }

    fun setImageBackground(@ColorInt backColor: Int): FMHelper {
        this.imageBackground = ColorDrawable(backColor)
        return this
    }


    fun setMainBtnBaseColor(@ColorInt color: Int): FMHelper {
        this.mainBtnBaseColor = color
        return this
    }

    fun setMainBtnBaseColor(color: String): FMHelper {
        return setMainBtnBaseColor(Color.parseColor(color))
    }


    fun setMainBtnOpenColor(@ColorInt color: Int): FMHelper {
        this.mainBtnOpenColor = color
        return this
    }

    fun setMainBtnOpenColor(color: String): FMHelper {
        return setMainBtnOpenColor(Color.parseColor(color))
    }


    fun setOptionBtnColors(@ColorInt optionBtnColors: Array<Int>): FMHelper {
        require(optionBtnColors.size >= numOptions) {
            ("Length of colors given is not equal to numOptions set in constructor ("
                    + numOptions + ", " + optionBtnColors.size + ")")
        }
        this.optionBtnColors = optionBtnColors
        return this
    }

    fun setOptionBtnColors(@ColorInt optionBtnColor: Int): FMHelper {
        Arrays.fill(optionBtnColors, optionBtnColor)
        return this
    }


    fun setOptionTexts(optionTexts: Array<String>): FMHelper {
        require(optionTexts.size >= numOptions) {
            ("Length of texts given is not equal to numOptions set in constructor ("
                    + numOptions + ", " + optionTexts.size + ")")
        }
        this.optionTexts = optionTexts
        return this
    }

    fun setOptionTexts(@StringRes optionTexts: IntArray): FMHelper {
        require(optionTexts.size == numOptions) {
            ("Length of texts given is not equal to numOptions set in constructor ("
                    + numOptions + ", " + optionTexts.size + ")")
        }
        for (i in optionTexts.indices)
            this.optionTexts?.set(i, resourceManager.getString(optionTexts[i]))
        return this
    }

    fun setOptionTextBackground(@DrawableRes optionTextBackground: Int): FMHelper {
        return setOptionTextBackground(resourceManager.getDrawable(optionTextBackground))
    }

    fun setOptionTextBackground(optionTextBackground: Drawable?): FMHelper {
        this.optionTextBackground = optionTextBackground
        return this
    }

    fun setOptionTextBackgroundColor(@ColorInt color: Int): FMHelper {
        return setOptionTextBackground(ColorDrawable(color))
    }


    fun setStatusBarColor(@ColorInt statusBarColor: Int): FMHelper {
        this.statusBarColor = statusBarColor
        return this
    }

    fun setStatusBarColor(color: String): FMHelper {
        return setStatusBarColor(Color.parseColor(color))
    }


    fun setAnimatedStatusBar(animatedStatusBar: Boolean): FMHelper {
        this.isAnimatedStatusBar = animatedStatusBar
        return this
    }


    fun setTintedFabSrc(tintedFabSrc: Boolean): FMHelper {
        isTintedFabSrc = tintedFabSrc
        return this
    }

    fun setWindow(window: Window): FMHelper {
        this.window = window
        return this
    }

    internal fun getUseCard(): Boolean {
        return useCard
    }

    internal fun setUseCard(useCard: Boolean): FMHelper {
        this.useCard = useCard
        return this
    }
}
